witness/internal/witness/witness.go

```diff
@@ -0,0 +1,175 @@
+
+	// SetCheckpoint adds a checkpoint to the storage for a given log.
+	SetCheckpoint(logPK string, c Chkpt) error
+}
+
+type Witness struct {
+	db     ChkptStorage
+	Signer note.Signer
+	Hasher hashers.LogHasher
+	LogV   logverifier.LogVerifier
+	SigVs  []note.Verifier
+}
+
+// NewWitness creates a new witness, which initially has no logs to follow.
+func NewWitness(h hashers.LogHasher, db ChkptStorage, sk string) (*Witness, error) {
+	ns, err := note.NewSigner(sk)
+	if err != nil {
+		return nil, err
+	}
+	return &Witness{
+		db:     db,
+		Signer: ns,
+		Hasher: h,
+		LogV:   logverifier.New(h),
+		SigVs:  []note.Verifier{},
+	}, nil
+}
+
+// parse verifies the checkpoint under the given list of possible public keys.
+// It returns the parsed checkpoint and the list of public keys under which the
+// signature(s) verified.
+func (w *Witness) parse(chkptRaw []byte, verifiers []note.Verifier) (*log.Checkpoint, []string, error) {
+	n, err := note.Open(chkptRaw, note.VerifierList(verifiers...))
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to verify checkpoint: %w", err)
+	}
+	chkpt := &log.Checkpoint{}
+	_, err = chkpt.Unmarshal([]byte(n.Text))
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to unmarshal new checkpoint: %w", err)
+	}
+	var keys []string
+	for _, sig := range n.Sigs {
+		keys = append(keys, sig.Name)
+	}
+	return chkpt, keys, nil
+}
+
+// GetLatest gets the latest checkpoint for a given log, which should be
+// consistent with all other checkpoints for the same log.  It also signs it
+// under the witness' key.
+func (w *Witness) GetLatest(logPK string) ([]byte, error) {
+	chkpt, err := w.db.GetLatest(logPK)
+	if err != nil {
+		return nil, err
+	}
+	// Add the witness' signature to the checkpoint.
+	// TODO find a way so we don't have to verify the signatures again.
+	n, err := note.Open(chkpt.Raw, note.VerifierList(w.SigVs...))
+	if err != nil {
+		return nil, fmt.Errorf("failed to verify checkpoint: %w", err)
+	}
+	cosigned, err := note.Sign(n, w.Signer)
+	if err != nil {
+		return nil, err
+	}
+	return cosigned, nil
+}
+
+// Update updates latest checkpoint if `to` matches the current version (`from`).
+func (w *Witness) Update(fromRaw, toRaw []byte, proof log.Proof) error {
+	// Check the signatures on both of the raw checkpoints and parse them
+	// into the log.Checkpoint format.
+	from, keysFrom, err := w.parse(fromRaw, w.SigVs)
+	if err != nil {
+		return err
+	}
+	to, keysTo, err := w.parse(toRaw, w.SigVs)
+	if err != nil {
+		return err
+	}
+	// Make sure both verified with respect to the same public key.
+	if len(keysFrom) > 1 || len(keysTo) > 1 {
+		return fmt.Errorf("Bad signature verification")
+	}
+	if keysFrom[0] != keysTo[0] {
+		return fmt.Errorf("Checkpoints verified under different public keys")
+	}
+	// Get the latest one because we don't want consistency proofs with
+	// respect to older checkpoints.
+	logPK := keysFrom[0]
+	latest, err := w.db.GetLatest(logPK)
```

Comment:
I think there is a risk of a race condition here. It's vanishingly small, in the context, but still....

Say the witness has checkpoint A tracked. The log then creates a split view, with checkpoints X1 and X2. Both of these are consistent with A, but inconsistent with each other. The following could happen:
```
w.Update(A, X1) // thread 1
w.Update(A, X2) // thread 2
```

If both threads pass this point then `latest = A` in both cases and consistency will pass. We'll eventually call `w.db.SetCheckpoint` with two inconsistent checkpoints. The following sequence is then possible:

```
w.db.SetCheckpoint(X1) // thread 1
w.GetLatest() -> X1 // any random thread
w.db.SetCheckpoint(X2) // thread 2
w.GetLatest() -> X2 // any random thread
```

The solution to this is enforce at write time the expected value of the existing checkpoint, or break the DB abstraction and open a transaction before the checkpoint read, and use DB transaction violation to enforce this.

Reply:
Okay, SetCheckpoint now requires the caller to say what it thinks the latest checkpoint is, and it writes the new one only if they're right.  Let me know what you think?

Reply:
The shape of this is now definitely improved. I've pointed out some peculiarities in the implementation of that method w.r.t. `nil` handling of the expected checkpoint, but I think this comment can be resolved now.

Reply:
I think so too :).  Thanks for your help!